src/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -589,7 +589,7 @@ public static void ExtensionProperty_InvalidDictionary()
             Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(obj1));
 
             ClassWithInvalidExtensionPropertyObjectString obj2 = new ClassWithInvalidExtensionPropertyObjectString();
-            Assert.Throws<NotSupportedException>(() => JsonSerializer.Serialize(obj2));
+            Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(obj2));
```

Comment:
Can you explain why the change here. Not sure which is more appropriate, but I believe the test object uses an invalid Key which will never be supported, so InvalidOperation may be more appropriate than NotSupported (based on past discussions).

Reply:
I believe it should be InvalidOperation, like the test above it. NSE was thrown before this change because the extension dictionary (`Dictionary<DummyObj, string>`)is not a type supported for (de)serialization. We were detecting this during warm up: https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L561.

The preceding test doesn't throw NSE because the extension dictionary (Dictionary<DummyObj, string>) is supported, just not as extension data. Instead, it throws InvalidOperation: https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L246-L250

Wherever possible, I think it's better to throw NSE for unsupported types, when the unsupported action (deserialization or serialization) is attempted, and not during warm up. This way, we don't block supported usage if the unsupported action isn't attempted (e.g. when serialization is supported (and attempted) but deserialization is not (but not attempted), or, when deserializing, there's missing JSON for the unsupported property or the property is ignored). 

Moreover, not throwing NSE during warm up/runtime information detection in this case allows the correct exception (InvalidOperation) to eventually be thrown, thus the change.